tools/build/mingw_toolchain/file_collector.cpp

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <algorithm>
+#include <direct.h>
+#include <iostream>
+#include <fstream>
+#include <shlwapi.h>
+#include <sstream>
+#include <string>
+#include <vector>
+#include <windows.h>
+
+#include "file_collector.h"
+
+namespace {
+
+std::string GetCWD() {
+  char* cwd = _getcwd(nullptr, 0);
+  if (cwd == nullptr) {
+    std::cerr << "Failed to get workding dir" << std::endl;
+  }
+  std::string result = R"(\\?\)" + std::string(cwd);
+  if (result.back() != '\\') {
+    result.push_back('\\');
+  }
+  delete cwd;
+  return result;
+}
+
+// Attempts to create a new folder in the temporary folder based off the PID.
+std::string GetTempDir(const std::string& baseName) {
+  TCHAR buffer[MAX_PATH + 1 + baseName.length() + 5 + 1];
```

Comment:
Weird mix of `TCHAR` and `char` functions here. I'd just stick with good old `char`, and use the `GetTempPathA` explicit variant.

`MAX_PATH + 1 + baseName.length() + 5 + 1` is a bit weird too. Technically correct, but not sure it is intentionally correct. I assume the `+ 1` is for a null terminator, not sure why we have two. `+ 5` I'm guessing is for `"%s_%04x\\"`, but I count 6 characters.

Reply:
Done. `GetTempPath` returns the length excluding the terminator. The docs say it's max return value is `MAX_PATH + 1`. Yes, it should have been `+ 6`, I missed the trailing `\`...